# Case Intro
Term deposits are a major source of income for a bank. A term deposit is a cash investment held at a financial institution. Your money is invested for an agreed rate of interest over a fixed amount of time, or term. The bank has various outreach plans to sell term deposits to their customers such as email marketing, advertisements, telephonic marketing, and digital marketing.

Telephonic marketing campaigns still remain one of the most effective way to reach out to people. However, they require huge investment as large call centers are hired to actually execute these campaigns. Hence, it is crucial to identify the customers most likely to convert beforehand so that they can be specifically targeted via call.

The data is related to direct marketing campaigns (phone calls) of a Portuguese banking institution. The classification goal is to predict if the client will subscribe to a term deposit (variable y).

Content
The data is related to the direct marketing campaigns of a Portuguese banking institution. The marketing campaigns were based on phone calls. Often, more than one contact to the same client was required, in order to access if the product (bank term deposit) would be ('yes') or not ('no') subscribed by the customer or not. The data folder contains two datasets:-

Bank.csv: 45,211 rows and 18 columns ordered by date (from May 2008 to November 2010)

Detailed Column Descriptions
bank client data:

1 - age (numeric)

2 - job : type of job (categorical: "admin.","unknown","unemployed","management","housemaid","entrepreneur","student",
"blue-collar","self-employed","retired","technician","services")

3 - marital : marital status (categorical: "married","divorced","single"; note: "divorced" means divorced or widowed)

4 - education (categorical: "unknown","secondary","primary","tertiary")

5 - default: has credit in default? (binary: "yes","no")

6 - balance: average yearly balance, in euros (numeric)

7 - housing: has housing loan? (binary: "yes","no")

8 - loan: has personal loan? (binary: "yes","no")
# related with the last contact of the current campaign:
9 - contact: contact communication type (categorical: "unknown","telephone","cellular")
10 - day: last contact day of the month (numeric)

11 - month: last contact month of year (categorical: "jan", "feb", "mar", …, "nov", "dec")

12 - duration: last contact duration, in seconds (numeric)

# other attributes:
13 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)

14 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric, -1 means client was not previously contacted)

15 - previous: number of contacts performed before this campaign and for this client (numeric)

16 - poutcome: outcome of the previous marketing campaign (categorical: "unknown","other","failure","success")

Output variable (desired target):

17 - y - has the client subscribed a term deposit? (binary: "yes","no")

Missing Attribute Values: None


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
df=pd.read_csv('https://raw.githubusercontent.com/ogut77/DataScience/main/data/Bank.csv',sep = ';')
df

,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,y
0,58,management,married,tertiary,no,2143,yes,no,unknown,5,may,261,1,-1,0,unknown,no
1,44,technician,single,secondary,no,29,yes,no,unknown,5,may,151,1,-1,0,unknown,no
2,33,entrepreneur,married,secondary,no,2,yes,yes,unknown,5,may,76,1,-1,0,unknown,no
3,47,blue-collar,married,unknown,no,1506,yes,no,unknown,5,may,92,1,-1,0,unknown,no
4,33,unknown,single,unknown,no,1,no,no,unknown,5,may,198,1,-1,0,unknown,no
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
45206,51,technician,married,tertiary,no,825,no,no,cellular,17,nov,977,3,-1,0,unknown,yes
45207,71,retired,divorced,primary,no,1729,no,no,cellular,17,nov,456,2,-1,0,unknown,yes
45208,72,retired,married,secondary,no,5715,no,no,cellular,17,nov,1127,5,184,3,success,yes
45209,57,blue-collar,married,secondary,no,668,no,no,telephone,17,nov,508,4,-1,0,unknown,no


In [2]:
print(df.shape)
df.info()
df.isnull().sum()

(45211, 17)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        45211 non-null  int64 
 1   job        45211 non-null  object
 2   marital    45211 non-null  object
 3   education  45211 non-null  object
 4   default    45211 non-null  object
 5   balance    45211 non-null  int64 
 6   housing    45211 non-null  object
 7   loan       45211 non-null  object
 8   contact    45211 non-null  object
 9   day        45211 non-null  int64 
 10  month      45211 non-null  object
 11  duration   45211 non-null  int64 
 12  campaign   45211 non-null  int64 
 13  pdays      45211 non-null  int64 
 14  previous   45211 non-null  int64 
 15  poutcome   45211 non-null  object
 16  y          45211 non-null  object
dtypes: int64(7), object(10)
memory usage: 5.9+ MB


,0
age,0
job,0
marital,0
education,0
default,0
balance,0
housing,0
loan,0
contact,0
day,0


In [3]:
#For object check the data
for cn in df.columns:
  if(df[cn].dtype==object):
    print(df[cn].value_counts())


job
blue-collar      9732
management       9458
technician       7597
admin.           5171
services         4154
retired          2264
self-employed    1579
entrepreneur     1487
unemployed       1303
housemaid        1240
student           938
unknown           288
Name: count, dtype: int64
marital
married     27214
single      12790
divorced     5207
Name: count, dtype: int64
education
secondary    23202
tertiary     13301
primary       6851
unknown       1857
Name: count, dtype: int64
default
no     44396
yes      815
Name: count, dtype: int64
housing
yes    25130
no     20081
Name: count, dtype: int64
loan
no     37967
yes     7244
Name: count, dtype: int64
contact
cellular     29285
unknown      13020
telephone     2906
Name: count, dtype: int64
month
may    13766
jul     6895
aug     6247
jun     5341
nov     3970
apr     2932
feb     2649
jan     1403
oct      738
sep      579
mar      477
dec      214
Name: count, dtype: int64
poutcome
unknown    36959
failure     4901
other  

In [4]:
# Following function converts non-numeric variables (e.g., 'category', 'object') into numeric using label encoding
# Note:
#Label encoding converts categorical values into integer codes. Each unique category is assigned a distinct number, such as 1, 2, 3, 4, etc.
#Example : In our data, education variable has following values:{ 'primary' 'secondary', 'tertiary', 'unknown'}
#and it is mapped { 'primary': 0, 'secondary': 1, 'tertiary': 2, 'unknown': 3}

#  education  education_encoded
#  secondary                  1
#   tertiary                  2
#    primary                  0
#    unknown                  3
#   tertiary                  2
#    primary                  0
#
#Label encoding is generally suitable for tree-based models (e.g., decision trees, random forests, boosting methods).
#However, it may not be appropriate for models where the objective function relies on distance-based calculations, such as neural networks, support vector machines (SVM), or linear regression.
#For nominal features (categories with no intrinsic ordering, e.g., "red," "blue," "green"), label encoding can mislead the model by implying an ordinal relationship where none exists.
#In such cases, one-hot encoding is usually preferred.
#One-hot encoding creates binary columns (dummy variables) for each category in a categorical feature, avoiding the introduction of unintended ordinality.

def Encoder(df):
          from sklearn import preprocessing
          columnsToEncode = list(df.select_dtypes(include=['category','object']))
          le = preprocessing.LabelEncoder()
          for feature in columnsToEncode:
              try:
                  df[feature] = le.fit_transform(df[feature])
              except:
                  print('Error encoding '+feature)
          return df


In [5]:
df=Encoder(df)
df

,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,y
0,58,4,1,2,0,2143,1,0,2,5,8,261,1,-1,0,3,0
1,44,9,2,1,0,29,1,0,2,5,8,151,1,-1,0,3,0
2,33,2,1,1,0,2,1,1,2,5,8,76,1,-1,0,3,0
3,47,1,1,3,0,1506,1,0,2,5,8,92,1,-1,0,3,0
4,33,11,2,3,0,1,0,0,2,5,8,198,1,-1,0,3,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
45206,51,9,1,2,0,825,0,0,0,17,9,977,3,-1,0,3,1
45207,71,5,0,0,0,1729,0,0,0,17,9,456,2,-1,0,3,1
45208,72,5,1,1,0,5715,0,0,0,17,9,1127,5,184,3,2,1
45209,57,1,1,1,0,668,0,0,1,17,9,508,4,-1,0,3,0


In [6]:
y = df['y'] #Output
X = df.drop('y',axis=1)
X

,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome
0,58,4,1,2,0,2143,1,0,2,5,8,261,1,-1,0,3
1,44,9,2,1,0,29,1,0,2,5,8,151,1,-1,0,3
2,33,2,1,1,0,2,1,1,2,5,8,76,1,-1,0,3
3,47,1,1,3,0,1506,1,0,2,5,8,92,1,-1,0,3
4,33,11,2,3,0,1,0,0,2,5,8,198,1,-1,0,3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
45206,51,9,1,2,0,825,0,0,0,17,9,977,3,-1,0,3
45207,71,5,0,0,0,1729,0,0,0,17,9,456,2,-1,0,3
45208,72,5,1,1,0,5715,0,0,0,17,9,1127,5,184,3,2
45209,57,1,1,1,0,668,0,0,1,17,9,508,4,-1,0,3


In [7]:
from sklearn.model_selection import  train_test_split


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=17)




Q1)Using  Random Forest,XGBoost, Light GBM and Gradient Boosting Classifier with default parameters (no parameter specifications except random_state) calculate Accuracy on Test data. Which method gives the best accuracy on test data

Q2) Using optuna hyperparmeter optimization technique and 100 trial

 a)find best methods with  parameters  using Cross validation (CV=3) technique for the range of   parameters below. What are the best parameters for the method with highest cross validation accuracy?
 For random forest


  "max_depth"   : trial.suggest_int("max_depth", 2,  X_train.shape[1]),
  "max_features": trial.suggest_int("max_features", 2, X_train.shape[1])

For XGBoost, Light GBM and Gradient Boosting Classifier

  "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
  "learning_rate": trial.suggest_float("learning_rate", 0.001, 0.3,log=True)

where X_train.shape[1] is number of columnns in the train data.

 b)Evaluate the performance of the  method with highest cross validation accuracy on test data.What is the accuracy value? Are there any improvement of the same method with default parameters?


For Q3 and Q4 ,use the following data.

In [8]:
dr=pd.read_csv('https://raw.githubusercontent.com/ogut77/DataScience/main/data/diamond.csv')
dr

,Carat Weight,Cut,Color,Clarity,Polish,Symmetry,Report,Price
0,1.10,Ideal,H,SI1,VG,EX,GIA,5169
1,0.83,Ideal,H,VS1,ID,ID,AGSL,3470
2,0.85,Ideal,H,SI1,EX,EX,GIA,3183
3,0.91,Ideal,E,SI1,VG,VG,GIA,4370
4,0.83,Ideal,G,SI1,EX,EX,GIA,3171
...,...,...,...,...,...,...,...,...
5995,1.03,Ideal,D,SI1,EX,EX,GIA,6250
5996,1.00,Very Good,D,SI1,VG,VG,GIA,5328
5997,1.02,Ideal,D,SI1,EX,EX,GIA,6157
5998,1.27,Signature-Ideal,G,VS1,EX,EX,GIA,11206


In [9]:
def Encoder(df):
          from sklearn import preprocessing
          columnsToEncode = list(df.select_dtypes(include=['category','object']))
          le = preprocessing.LabelEncoder()
          for feature in columnsToEncode:
              try:
                  df[feature] = le.fit_transform(df[feature])
              except:
                  print('Error encoding '+feature)
          return df


In [10]:
dr=Encoder(dr)
dr

,Carat Weight,Cut,Color,Clarity,Polish,Symmetry,Report,Price
0,1.10,2,4,2,3,0,1,5169
1,0.83,2,4,3,2,2,0,3470
2,0.85,2,4,2,0,0,1,3183
3,0.91,2,1,2,3,3,1,4370
4,0.83,2,3,2,0,0,1,3171
...,...,...,...,...,...,...,...,...
5995,1.03,2,0,2,0,0,1,6250
5996,1.00,4,0,2,3,3,1,5328
5997,1.02,2,0,2,0,0,1,6157
5998,1.27,3,3,3,0,0,1,11206


In [11]:
y = dr['Price'] #Output
X = dr.drop('Price',axis=1)
X

,Carat Weight,Cut,Color,Clarity,Polish,Symmetry,Report
0,1.10,2,4,2,3,0,1
1,0.83,2,4,3,2,2,0
2,0.85,2,4,2,0,0,1
3,0.91,2,1,2,3,3,1
4,0.83,2,3,2,0,0,1
...,...,...,...,...,...,...,...
5995,1.03,2,0,2,0,0,1
5996,1.00,4,0,2,3,3,1
5997,1.02,2,0,2,0,0,1
5998,1.27,3,3,3,0,0,1


In [12]:
from sklearn.model_selection import  train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=17)

In [6]:
#QUESTION 1

# ==============================
# 1. Import Libraries
# ==============================
import pandas as pd
import numpy as np

from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

# Install if needed:
# !pip install xgboost lightgbm

from xgboost import XGBClassifier
from lightgbm import LGBMClassifier


# ==============================
# 2. Load Data
# ==============================
df = pd.read_csv(
    'https://raw.githubusercontent.com/ogut77/DataScience/main/data/Bank.csv',
    sep=';'
)


# ==============================
# 3. Encode Categorical Variables
# ==============================
from sklearn.preprocessing import LabelEncoder

def Encoder(df):
    df = df.copy()
    cols = df.select_dtypes(include=['object']).columns
    le = LabelEncoder()

    for col in cols:
        df[col] = le.fit_transform(df[col])

    return df

df = Encoder(df)


# ==============================
# 4. Split Features & Target
# ==============================
y = df['y']   # already encoded to 0/1
X = df.drop('y', axis=1)


# ==============================
# 5. Train-Test Split
# ==============================
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=17
)


# ==============================
# 6. Initialize Models
# ==============================
models = {
    "Random Forest": RandomForestClassifier(random_state=17),
    "XGBoost": XGBClassifier(use_label_encoder=False, eval_metric='logloss', random_state=17),
    "LightGBM": LGBMClassifier(random_state=17),
    "Gradient Boosting": GradientBoostingClassifier(random_state=17)
}


# ==============================
# 7. Train & Evaluate Models
# ==============================
results = {}

for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    acc = accuracy_score(y_test, y_pred)
    results[name] = acc

    print(f"{name} Accuracy: {acc:.4f}")


# ==============================
# 8. Find Best Model
# ==============================
best_model = max(results, key=results.get)

print("\nBest Model:", best_model)
print("Best Accuracy:", results[best_model])

Random Forest Accuracy: 0.9025


/usr/local/lib/python3.12/dist-packages/xgboost/training.py:200: UserWarning: [18:04:14] WARNING: /__w/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


XGBoost Accuracy: 0.9046
[LightGBM] [Info] Number of positive: 3940, number of negative: 29968
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.004599 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 988
[LightGBM] [Info] Number of data points in the train set: 33908, number of used features: 16
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.116197 -> initscore=-2.028949
[LightGBM] [Info] Start training from score -2.028949
LightGBM Accuracy: 0.9074
Gradient Boosting Accuracy: 0.9019

Best Model: LightGBM
Best Accuracy: 0.9073697248518092


# QUESTION 1 ANSWER:
four classification models (Random Forest, XGBoost, LightGBM, and Gradient Boosting) were trained using default parameters, and their performance was evaluated using accuracy on the test dataset.

The obtained accuracies are:

Random Forest: 0.9025
XGBoost: 0.9046
LightGBM: 0.9074
Gradient Boosting: 0.9019

Among these models, LightGBM achieved the highest test accuracy of 0.9074, making it the best-performing model under default parameter settings.




In [8]:
#QUESTION 2

# =========================================
# QUESTION 2
# Optuna hyperparameter tuning (100 trials)
# =========================================

# Install if needed:
!pip install optuna xgboost lightgbm

import pandas as pd
import numpy as np
import optuna

from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier

from xgboost import XGBClassifier
from lightgbm import LGBMClassifier


# =========================================
# 1. Load Data
# =========================================
df = pd.read_csv(
    'https://raw.githubusercontent.com/ogut77/DataScience/main/data/Bank.csv',
    sep=';'
)


# =========================================
# 2. Encode categorical variables
# =========================================
def Encoder(df):
    df = df.copy()
    cols = df.select_dtypes(include=['object']).columns
    le = LabelEncoder()
    for col in cols:
        df[col] = le.fit_transform(df[col])
    return df

df = Encoder(df)


# =========================================
# 3. Split features and target
# =========================================
y = df['y']
X = df.drop('y', axis=1)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=17
)


# =========================================
# 4. Default model accuracies (needed for Q2b comparison)
# =========================================
default_models = {
    "Random Forest": RandomForestClassifier(random_state=17),
    "XGBoost": XGBClassifier(use_label_encoder=False, eval_metric='logloss', random_state=17),
    "LightGBM": LGBMClassifier(random_state=17),
    "Gradient Boosting": GradientBoostingClassifier(random_state=17)
}

default_results = {}

print("===== DEFAULT MODEL TEST ACCURACIES =====")
for name, model in default_models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    acc = accuracy_score(y_test, y_pred)
    default_results[name] = acc
    print(f"{name}: {acc:.4f}")


# =========================================
# 5. Optuna objective functions
# =========================================

# Random Forest:
# "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1])
# "max_features": trial.suggest_int("max_features", 2, X_train.shape[1])

def objective_rf(trial):
    params = {
        "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
        "max_features": trial.suggest_int("max_features", 2, X_train.shape[1]),
        "random_state": 17,
        "n_jobs": -1
    }
    model = RandomForestClassifier(**params)
    score = cross_val_score(model, X_train, y_train, cv=3, scoring='accuracy', n_jobs=-1).mean()
    return score


# XGBoost:
# "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1])
# "learning_rate": trial.suggest_float("learning_rate", 0.001, 0.3, log=True)

def objective_xgb(trial):
    params = {
        "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
        "learning_rate": trial.suggest_float("learning_rate", 0.001, 0.3, log=True),
        "random_state": 17,
        "eval_metric": "logloss"
    }
    model = XGBClassifier(**params)
    score = cross_val_score(model, X_train, y_train, cv=3, scoring='accuracy', n_jobs=-1).mean()
    return score


# LightGBM:
def objective_lgb(trial):
    params = {
        "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
        "learning_rate": trial.suggest_float("learning_rate", 0.001, 0.3, log=True),
        "random_state": 17
    }
    model = LGBMClassifier(**params)
    score = cross_val_score(model, X_train, y_train, cv=3, scoring='accuracy', n_jobs=-1).mean()
    return score


# Gradient Boosting:
def objective_gb(trial):
    params = {
        "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
        "learning_rate": trial.suggest_float("learning_rate", 0.001, 0.3, log=True),
        "random_state": 17
    }
    model = GradientBoostingClassifier(**params)
    score = cross_val_score(model, X_train, y_train, cv=3, scoring='accuracy', n_jobs=-1).mean()
    return score


# =========================================
# 6. Run Optuna studies (100 trials each)
# =========================================
print("\n===== OPTUNA TUNING STARTED =====")

study_rf = optuna.create_study(direction='maximize')
study_rf.optimize(objective_rf, n_trials=100)

study_xgb = optuna.create_study(direction='maximize')
study_xgb.optimize(objective_xgb, n_trials=100)

study_lgb = optuna.create_study(direction='maximize')
study_lgb.optimize(objective_lgb, n_trials=100)

study_gb = optuna.create_study(direction='maximize')
study_gb.optimize(objective_gb, n_trials=100)


# =========================================
# 7. Print best parameters and CV scores
# =========================================
print("\n===== BEST PARAMETERS AND CV ACCURACIES =====")
print("Random Forest Best Params:", study_rf.best_params)
print("Random Forest Best CV Accuracy:", round(study_rf.best_value, 4))

print("\nXGBoost Best Params:", study_xgb.best_params)
print("XGBoost Best CV Accuracy:", round(study_xgb.best_value, 4))

print("\nLightGBM Best Params:", study_lgb.best_params)
print("LightGBM Best CV Accuracy:", round(study_lgb.best_value, 4))

print("\nGradient Boosting Best Params:", study_gb.best_params)
print("Gradient Boosting Best CV Accuracy:", round(study_gb.best_value, 4))


# =========================================
# 8. Find overall best method from CV
# =========================================
studies = {
    "Random Forest": study_rf,
    "XGBoost": study_xgb,
    "LightGBM": study_lgb,
    "Gradient Boosting": study_gb
}

best_method = max(studies, key=lambda k: studies[k].best_value)
best_study = studies[best_method]
best_params = best_study.best_params

print("\n===== BEST METHOD FROM CROSS VALIDATION =====")
print("Best Method:", best_method)
print("Best Parameters:", best_params)
print("Best CV Accuracy:", round(best_study.best_value, 4))


# =========================================
# 9. Train tuned best model on training data
#    and evaluate on test data
# =========================================
if best_method == "Random Forest":
    tuned_model = RandomForestClassifier(
        **best_params,
        random_state=17,
        n_jobs=-1
    )

elif best_method == "XGBoost":
    tuned_model = XGBClassifier(
        **best_params,
        random_state=17,
        eval_metric='logloss'
    )

elif best_method == "LightGBM":
    tuned_model = LGBMClassifier(
        **best_params,
        random_state=17
    )

elif best_method == "Gradient Boosting":
    tuned_model = GradientBoostingClassifier(
        **best_params,
        random_state=17
    )

tuned_model.fit(X_train, y_train)
y_pred_tuned = tuned_model.predict(X_test)
tuned_test_acc = accuracy_score(y_test, y_pred_tuned)

print("\n===== TEST PERFORMANCE OF BEST TUNED METHOD =====")
print("Best Tuned Method:", best_method)
print("Tuned Test Accuracy:", round(tuned_test_acc, 4))


# =========================================
# 10. Compare with default version of same method
# =========================================
default_same_method_acc = default_results[best_method]

print("\n===== COMPARISON WITH DEFAULT VERSION =====")
print(f"Default {best_method} Test Accuracy:", round(default_same_method_acc, 4))
print(f"Tuned   {best_method} Test Accuracy:", round(tuned_test_acc, 4))

if tuned_test_acc > default_same_method_acc:
    print("Improvement: Yes")
elif tuned_test_acc < default_same_method_acc:
    print("Improvement: No, tuned model is worse")
else:
    print("Improvement: No change")

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 419.5/419.5 kB 17.8 MB/s eta 0:00:00
===== DEFAULT MODEL TEST ACCURACIES =====
Random Forest: 0.9025


/usr/local/lib/python3.12/dist-packages/xgboost/training.py:200: UserWarning: [18:07:23] WARNING: /__w/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


XGBoost: 0.9046
[LightGBM] [Info] Number of positive: 3940, number of negative: 29968
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.004050 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 988
[LightGBM] [Info] Number of data points in the train set: 33908, number of used features: 16
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.116197 -> initscore=-2.028949
[LightGBM] [Info] Start training from score -2.028949
LightGBM: 0.9074


[I 2026-04-18 18:07:30,394] A new study created in memory with name: no-name-0e4bbd65-df54-4c30-ac4b-aa67432bb8fc


Gradient Boosting: 0.9019

===== OPTUNA TUNING STARTED =====


[I 2026-04-18 18:07:54,691] Trial 0 finished with value: 0.9046538488757833 and parameters: {'max_depth': 12, 'max_features': 15}. Best is trial 0 with value: 0.9046538488757833.
[I 2026-04-18 18:07:58,169] Trial 1 finished with value: 0.8867818610501437 and parameters: {'max_depth': 4, 'max_features': 3}. Best is trial 0 with value: 0.9046538488757833.
[I 2026-04-18 18:08:02,043] Trial 2 finished with value: 0.893329014676404 and parameters: {'max_depth': 5, 'max_features': 4}. Best is trial 0 with value: 0.9046538488757833.
[I 2026-04-18 18:08:15,155] Trial 3 finished with value: 0.9058334636565325 and parameters: {'max_depth': 11, 'max_features': 9}. Best is trial 3 with value: 0.9058334636565325.
[I 2026-04-18 18:08:20,479] Trial 4 finished with value: 0.8962486613919601 and parameters: {'max_depth': 4, 'max_features': 10}. Best is trial 3 with value: 0.9058334636565325.
[I 2026-04-18 18:08:29,754] Trial 5 finished with value: 0.9046243399147581 and parameters: {'max_depth': 7, 'ma


===== BEST PARAMETERS AND CV ACCURACIES =====
Random Forest Best Params: {'max_depth': 14, 'max_features': 10}
Random Forest Best CV Accuracy: 0.906

XGBoost Best Params: {'max_depth': 4, 'learning_rate': 0.14080627483532862}
XGBoost Best CV Accuracy: 0.9081

LightGBM Best Params: {'max_depth': 13, 'learning_rate': 0.06726993554721772}
LightGBM Best CV Accuracy: 0.9094

Gradient Boosting Best Params: {'max_depth': 4, 'learning_rate': 0.13752927779631416}
Gradient Boosting Best CV Accuracy: 0.907

===== BEST METHOD FROM CROSS VALIDATION =====
Best Method: LightGBM
Best Parameters: {'max_depth': 13, 'learning_rate': 0.06726993554721772}
Best CV Accuracy: 0.9094
[LightGBM] [Info] Number of positive: 3940, number of negative: 29968
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.005992 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 988
[L

# QUESTION 2 ANSWER:
Q2(a)

Optuna hyperparameter optimization was applied to all four models using 100 trials and 3-fold cross-validation.

The best parameters and corresponding cross-validation accuracies are:

Random Forest:
Best CV Accuracy = 0.906
Parameters = {max_depth: 14, max_features: 10}
XGBoost:
Best CV Accuracy = 0.9081
Parameters = {max_depth: 4, learning_rate: 0.1408}
LightGBM:
Best CV Accuracy = 0.9094
Parameters = {max_depth: 13, learning_rate: 0.0673}
Gradient Boosting:
Best CV Accuracy = 0.907
Parameters = {max_depth: 4, learning_rate: 0.1375}

Among all models, LightGBM achieved the highest cross-validation accuracy (0.9094) and was selected as the best model.

Q2(b)

The best model (LightGBM with optimized parameters) was evaluated on the test dataset.

Tuned LightGBM Test Accuracy: 0.9061
Default LightGBM Test Accuracy: 0.9074

Although hyperparameter tuning improved the cross-validation performance, the tuned model performed slightly worse on the test data compared to the default model.

Q3)Using Linear Regression,Decison Tree Random Forest,XGBoost, Light GBM and Gradient Boosting Classifier with default parameters (no parameter specifications except random_state) calculate R2 statistics on test data. Which method gives the best accuracy on test data

Q4) Using optuna hyperparmeter optimization technique (100 trial)  with Random Forest,XGBoost, Light GBM and Gradient Boosting Regressor


a)find best methods with  parameters  using Cross validation (CV=3) technique for the range of   parameters below. What are the best parameters for the method with highest cross validation R2?

For random forest


  "max_depth"   : trial.suggest_int("max_depth", 2,  X_train.shape[1]),
  "max_features": trial.suggest_int("max_features", 2, X_train.shape[1])

For XGBoost, Light GBM and Gradient Boosting Classifier

  "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
  "learning_rate": trial.suggest_float("learning_rate", 0.001, 0.3,log=True)

where X_train.shape[1] is number of columnns in the train data.

 b)Evaluate the performance of the  method with highest cross validation R2 on test data. What is the R2 value? Are there any improvement of the same method with default parameters?


In [9]:
#QUESTION 3

# =========================================
# QUESTION 3
# Default models - compare R2 on test data
# =========================================

# Install if needed:
# !pip install xgboost lightgbm

import pandas as pd
import numpy as np

from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor

from xgboost import XGBRegressor
from lightgbm import LGBMRegressor


# =========================================
# 1. Load Data
# =========================================
dr = pd.read_csv('https://raw.githubusercontent.com/ogut77/DataScience/main/data/diamond.csv')


# =========================================
# 2. Encode categorical variables
# =========================================
def Encoder(df):
    df = df.copy()
    cols = df.select_dtypes(include=['object']).columns
    le = LabelEncoder()
    for col in cols:
        df[col] = le.fit_transform(df[col])
    return df

dr = Encoder(dr)


# =========================================
# 3. Split features and target
# =========================================
y = dr['Price']
X = dr.drop('Price', axis=1)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=17
)


# =========================================
# 4. Initialize default models
# =========================================
models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(random_state=17),
    "Random Forest": RandomForestRegressor(random_state=17, n_jobs=-1),
    "XGBoost": XGBRegressor(random_state=17, n_jobs=-1),
    "LightGBM": LGBMRegressor(random_state=17),
    "Gradient Boosting": GradientBoostingRegressor(random_state=17)
}


# =========================================
# 5. Train and evaluate with R2
# =========================================
results = {}

print("===== Q3: DEFAULT MODEL TEST R2 SCORES =====")
for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    r2 = r2_score(y_test, y_pred)
    results[name] = r2
    print(f"{name}: {r2:.4f}")


# =========================================
# 6. Best default model
# =========================================
best_model = max(results, key=results.get)
print("\nBest Default Model:", best_model)
print("Best Default Test R2:", round(results[best_model], 4))

===== Q3: DEFAULT MODEL TEST R2 SCORES =====
Linear Regression: 0.8234
Decision Tree: 0.9592
Random Forest: 0.9800
XGBoost: 0.9809
[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000215 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 193
[LightGBM] [Info] Number of data points in the train set: 4500, number of used features: 7
[LightGBM] [Info] Start training from score 11827.946667
LightGBM: 0.9813
Gradient Boosting: 0.9739

Best Default Model: LightGBM
Best Default Test R2: 0.9813


In [10]:
#QUESTION 4

# =========================================
# QUESTION 4
# Optuna tuning with 100 trials using CV=3
# Compare tuned vs default using R2
# =========================================

# Install if needed:
# !pip install optuna xgboost lightgbm

import pandas as pd
import numpy as np
import optuna

from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import r2_score

from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor

from xgboost import XGBRegressor
from lightgbm import LGBMRegressor


# =========================================
# 1. Load Data
# =========================================
dr = pd.read_csv('https://raw.githubusercontent.com/ogut77/DataScience/main/data/diamond.csv')


# =========================================
# 2. Encode categorical variables
# =========================================
def Encoder(df):
    df = df.copy()
    cols = df.select_dtypes(include=['object']).columns
    le = LabelEncoder()
    for col in cols:
        df[col] = le.fit_transform(df[col])
    return df

dr = Encoder(dr)


# =========================================
# 3. Split features and target
# =========================================
y = dr['Price']
X = dr.drop('Price', axis=1)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=17
)


# =========================================
# 4. Default model R2 scores for comparison
# =========================================
default_models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(random_state=17),
    "Random Forest": RandomForestRegressor(random_state=17, n_jobs=-1),
    "XGBoost": XGBRegressor(random_state=17, n_jobs=-1),
    "LightGBM": LGBMRegressor(random_state=17),
    "Gradient Boosting": GradientBoostingRegressor(random_state=17)
}

default_results = {}

print("===== DEFAULT MODEL TEST R2 SCORES =====")
for name, model in default_models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    r2 = r2_score(y_test, y_pred)
    default_results[name] = r2
    print(f"{name}: {r2:.4f}")


# =========================================
# 5. Optuna objective functions
# =========================================

# Random Forest:
# "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1])
# "max_features": trial.suggest_int("max_features", 2, X_train.shape[1])

def objective_rf(trial):
    params = {
        "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
        "max_features": trial.suggest_int("max_features", 2, X_train.shape[1]),
        "random_state": 17,
        "n_jobs": -1
    }
    model = RandomForestRegressor(**params)
    score = cross_val_score(model, X_train, y_train, cv=3, scoring='r2', n_jobs=-1).mean()
    return score


# XGBoost:
# "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1])
# "learning_rate": trial.suggest_float("learning_rate", 0.001, 0.3, log=True)

def objective_xgb(trial):
    params = {
        "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
        "learning_rate": trial.suggest_float("learning_rate", 0.001, 0.3, log=True),
        "random_state": 17,
        "n_jobs": -1
    }
    model = XGBRegressor(**params)
    score = cross_val_score(model, X_train, y_train, cv=3, scoring='r2', n_jobs=-1).mean()
    return score


# LightGBM
def objective_lgb(trial):
    params = {
        "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
        "learning_rate": trial.suggest_float("learning_rate", 0.001, 0.3, log=True),
        "random_state": 17
    }
    model = LGBMRegressor(**params)
    score = cross_val_score(model, X_train, y_train, cv=3, scoring='r2', n_jobs=-1).mean()
    return score


# Gradient Boosting
def objective_gb(trial):
    params = {
        "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
        "learning_rate": trial.suggest_float("learning_rate", 0.001, 0.3, log=True),
        "random_state": 17
    }
    model = GradientBoostingRegressor(**params)
    score = cross_val_score(model, X_train, y_train, cv=3, scoring='r2', n_jobs=-1).mean()
    return score


# =========================================
# 6. Run Optuna studies
# =========================================
print("\n===== OPTUNA TUNING STARTED =====")

study_rf = optuna.create_study(direction='maximize')
study_rf.optimize(objective_rf, n_trials=100)

study_xgb = optuna.create_study(direction='maximize')
study_xgb.optimize(objective_xgb, n_trials=100)

study_lgb = optuna.create_study(direction='maximize')
study_lgb.optimize(objective_lgb, n_trials=100)

study_gb = optuna.create_study(direction='maximize')
study_gb.optimize(objective_gb, n_trials=100)


# =========================================
# 7. Print best parameters and CV R2
# =========================================
print("\n===== BEST PARAMETERS AND CV R2 SCORES =====")
print("Random Forest Best Params:", study_rf.best_params)
print("Random Forest Best CV R2:", round(study_rf.best_value, 4))

print("\nXGBoost Best Params:", study_xgb.best_params)
print("XGBoost Best CV R2:", round(study_xgb.best_value, 4))

print("\nLightGBM Best Params:", study_lgb.best_params)
print("LightGBM Best CV R2:", round(study_lgb.best_value, 4))

print("\nGradient Boosting Best Params:", study_gb.best_params)
print("Gradient Boosting Best CV R2:", round(study_gb.best_value, 4))


# =========================================
# 8. Find overall best tuned method by CV R2
# =========================================
studies = {
    "Random Forest": study_rf,
    "XGBoost": study_xgb,
    "LightGBM": study_lgb,
    "Gradient Boosting": study_gb
}

best_method = max(studies, key=lambda k: studies[k].best_value)
best_study = studies[best_method]
best_params = best_study.best_params

print("\n===== BEST METHOD FROM CROSS VALIDATION =====")
print("Best Method:", best_method)
print("Best Parameters:", best_params)
print("Best CV R2:", round(best_study.best_value, 4))


# =========================================
# 9. Train best tuned method and test it
# =========================================
if best_method == "Random Forest":
    tuned_model = RandomForestRegressor(
        **best_params,
        random_state=17,
        n_jobs=-1
    )

elif best_method == "XGBoost":
    tuned_model = XGBRegressor(
        **best_params,
        random_state=17,
        n_jobs=-1
    )

elif best_method == "LightGBM":
    tuned_model = LGBMRegressor(
        **best_params,
        random_state=17
    )

elif best_method == "Gradient Boosting":
    tuned_model = GradientBoostingRegressor(
        **best_params,
        random_state=17
    )

tuned_model.fit(X_train, y_train)
y_pred_tuned = tuned_model.predict(X_test)
tuned_test_r2 = r2_score(y_test, y_pred_tuned)

print("\n===== TEST PERFORMANCE OF BEST TUNED METHOD =====")
print("Best Tuned Method:", best_method)
print("Tuned Test R2:", round(tuned_test_r2, 4))


# =========================================
# 10. Compare with default version of same method
# =========================================
default_same_method_r2 = default_results[best_method]

print("\n===== COMPARISON WITH DEFAULT VERSION =====")
print(f"Default {best_method} Test R2:", round(default_same_method_r2, 4))
print(f"Tuned   {best_method} Test R2:", round(tuned_test_r2, 4))

if tuned_test_r2 > default_same_method_r2:
    print("Improvement: Yes")
elif tuned_test_r2 < default_same_method_r2:
    print("Improvement: No, tuned model is worse")
else:
    print("Improvement: No change")

===== DEFAULT MODEL TEST R2 SCORES =====
Linear Regression: 0.8234
Decision Tree: 0.9592
Random Forest: 0.9800
XGBoost: 0.9809
[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000208 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 193
[LightGBM] [Info] Number of data points in the train set: 4500, number of used features: 7
[LightGBM] [Info] Start training from score 11827.946667
LightGBM: 0.9813


[I 2026-04-18 19:14:39,506] A new study created in memory with name: no-name-89e38da5-1dae-43e7-97ec-0d87fb2eda9c


Gradient Boosting: 0.9739

===== OPTUNA TUNING STARTED =====


[I 2026-04-18 19:14:40,812] Trial 0 finished with value: 0.9710219510219829 and parameters: {'max_depth': 7, 'max_features': 7}. Best is trial 0 with value: 0.9710219510219829.
[I 2026-04-18 19:14:41,501] Trial 1 finished with value: 0.8938298760966701 and parameters: {'max_depth': 6, 'max_features': 2}. Best is trial 0 with value: 0.9710219510219829.
[I 2026-04-18 19:14:42,299] Trial 2 finished with value: 0.9153488693549434 and parameters: {'max_depth': 4, 'max_features': 6}. Best is trial 0 with value: 0.9710219510219829.
[I 2026-04-18 19:14:43,048] Trial 3 finished with value: 0.9375844290578602 and parameters: {'max_depth': 6, 'max_features': 3}. Best is trial 0 with value: 0.9710219510219829.
[I 2026-04-18 19:14:43,868] Trial 4 finished with value: 0.9073338809892725 and parameters: {'max_depth': 4, 'max_features': 7}. Best is trial 0 with value: 0.9710219510219829.
[I 2026-04-18 19:14:44,659] Trial 5 finished with value: 0.9187436916750902 and parameters: {'max_depth': 4, 'max_f


===== BEST PARAMETERS AND CV R2 SCORES =====
Random Forest Best Params: {'max_depth': 7, 'max_features': 5}
Random Forest Best CV R2: 0.9721

XGBoost Best Params: {'max_depth': 5, 'learning_rate': 0.13130912923439655}
XGBoost Best CV R2: 0.986

LightGBM Best Params: {'max_depth': 4, 'learning_rate': 0.29506953359019905}
LightGBM Best CV R2: 0.9837

Gradient Boosting Best Params: {'max_depth': 4, 'learning_rate': 0.17777192437249878}
Gradient Boosting Best CV R2: 0.987

===== BEST METHOD FROM CROSS VALIDATION =====
Best Method: Gradient Boosting
Best Parameters: {'max_depth': 4, 'learning_rate': 0.17777192437249878}
Best CV R2: 0.987

===== TEST PERFORMANCE OF BEST TUNED METHOD =====
Best Tuned Method: Gradient Boosting
Tuned Test R2: 0.9842

===== COMPARISON WITH DEFAULT VERSION =====
Default Gradient Boosting Test R2: 0.9739
Tuned   Gradient Boosting Test R2: 0.9842
Improvement: Yes


# In Question 3, R² scores were calculated for all models using default parameters. Among them, LightGBM achieved the highest test R² score of 0.9813, making it the best-performing method on the test data.

# In Question 4(a), hyperparameter optimization using Optuna with 100 trials and 3-fold cross-validation showed that Gradient Boosting achieved the highest cross-validation R² score of 0.987 with parameters max_depth = 4 and learning_rate ≈ 0.178.

# In Question 4(b), the tuned Gradient Boosting model achieved a test R² score of 0.9842, which is higher than its default test R² of 0.9739. Therefore, there is an improvement after hyperparameter optimization.